### PR TITLE
Refactor ArrayBufferView

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

File: std/assembly/typedarray.ts [hunk 1]
```diff
@@ -1593,12 +1549,12 @@ function FILTER<TArray extends ArrayBufferView, T>(
...
   // shrink output buffer
-  var length = j << alignof<T>();
-  var data = __realloc(buffer, length);
-  out.data = changetype<ArrayBuffer>(data); // retain
-  out.dataStart = data;
-  out.dataLength = length;
-  return out;
+  var byteLength = j << alignof<T>();
+  var data = __realloc(buffer, byteLength);
+  store<ArrayBuffer>(out, changetype<ArrayBuffer>(data), offsetof<TArray>("buffer"));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: appears to miss a `__retain`

File: std/assembly/typedarray.ts [hunk 2]
```diff
@@ -1741,9 +1697,9 @@ function WRAP<TArray extends ArrayBufferView, T>(buffer: ArrayBuffer, byteOffset
   if (byteOffset + byteLength > buffer.byteLength) {
     throw new RangeError(E_INVALIDLENGTH);
   }
-  var out = changetype<TArray>(__alloc(offsetof<TArray>(), idof<TArray>()));
-  out.data = buffer;
-  out.dataLength = byteLength;
-  out.dataStart = changetype<usize>(buffer) + <usize>byteOffset;
-  return out;
+  var out = __alloc(offsetof<TArray>(), idof<TArray>());
+  store<ArrayBuffer>(out, buffer, offsetof<TArray>("buffer"));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: as well

File: std/assembly/typedarray.ts [hunk 3]
```diff
@@ -1557,19 +1514,18 @@ function MAP<TArray extends ArrayBufferView, T>(
...
-  out.dataLength = byteLength;
-
   for (let i = 0; i < len; i++) {
     store<T>(
       buffer + (<usize>i << alignof<T>()),
       fn(load<T>(dataStart + (<usize>i << alignof<T>())), i, array)
     );
   }
-  return out;
+  store<ArrayBuffer>(out, changetype<ArrayBuffer>(buffer), offsetof<TArray>("buffer"));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: as well

File: std/assembly/typedarray.ts [hunk 4]
```diff
@@ -1487,11 +1443,12 @@ function SUBARRAY<TArray extends ArrayBufferView, T>(
...
   end   = end   < 0 ? max(len + end,   0) : min(end,   len);
   end   = max(end, begin);
-  var out = changetype<TArray>(__alloc(offsetof<TArray>(), idof<TArray>())); // retains
-  out.data = array.data; // retains
-  out.dataStart = array.dataStart + (<usize>begin << alignof<T>());
-  out.dataLength = (end - begin) << alignof<T>();
-  return out;
+
+  var out = __alloc(offsetof<TArray>(), idof<TArray>());
+  store<ArrayBuffer>(out, array.buffer, offsetof<TArray>("buffer"));
```
Comment (dcodeIO) [inline std/assembly/typedarray.ts]: as well

File: std/assembly/typedarray.ts (hunk 1 above)
Comment (dcodeIO) [reply std/assembly/typedarray.ts]: essentially, whenever a managed reference is assigned to a field, the compiler would first `__retain` the new value and then `__release` the old value, with the second step only relevant if the previous value isn't `null` (like when allocating a new array).

Comment (MaxGraey) [reply std/assembly/typedarray.ts]: Yeah. thanks!

File: std/assembly/dataview.ts [hunk 5]
```diff
@@ -6,9 +6,14 @@ import { E_INDEXOUTOFRANGE, E_INVALIDLENGTH } from "./util/error";
 
 export class DataView {
 
-  private data: ArrayBuffer;
-  private dataStart: usize;
-  private dataLength: i32;
+  @unsafe readonly dataStart: usize;
```
Comment (dcodeIO) [inline std/assembly/dataview.ts]: You mentioned that DataView should extend ArrayBufferView. Is this possible now?

File: std/assembly/index.d.ts [hunk 6]
```diff
@@ -1165,12 +1164,12 @@ interface ArrayBufferView<T> {
...
 
 /* @internal */
-declare abstract class TypedArray<T> implements ArrayBufferView<T> {
+declare abstract class TypedArray<T> implements ArrayBufferView {
   [key: number]: T;
   /** Number of bytes per element. */
   static readonly BYTES_PER_ELEMENT: usize;
   /** Wrap an ArrayBuffer */
-  static wrap(buffer: ArrayBuffer, byteOffset?: i32, length?: i32): TypedArray<T>;
+  static wrap<T>(buffer: ArrayBuffer, byteOffset?: i32, length?: i32): TypedArray<T>;
```
Comment (dcodeIO) [inline std/assembly/index.d.ts]: I see why this change was made, but it doesn't match the actual signature, which is non-generic on the concrete implementations, now.

File: std/assembly/arraybuffer.ts [hunk 7]
```diff
@@ -6,31 +6,28 @@ import { E_INVALIDLENGTH } from "./util/error";
 
 export abstract class ArrayBufferView {
 
-  @unsafe data: ArrayBuffer;
-  @unsafe dataStart: usize;
-  @unsafe dataLength: u32;
+  @unsafe readonly dataStart: usize;
```
Comment (dcodeIO) [inline std/assembly/arraybuffer.ts]: Wondering if there is a particular reason for changing the order here? Asking because if it remains this way, there's a constant in the loader that needs to be updated.

Comment (MaxGraey) [reply std/assembly/arraybuffer.ts]: the main reason is `dataStart` used more frequent then `buffer` so I guess have pure `load` without `offset=4` will be better. I'll fix loader

File: std/assembly/index.d.ts (hunk 6 above)
Comment (MaxGraey) [reply std/assembly/index.d.ts]: what if I change it to:
```
static wrap(buffer: ArrayBuffer, byteOffset?: i32, length?: i32): ArrayBufferView;
```
?

File: std/assembly/dataview.ts (hunk 5 above)
Comment (MaxGraey) [reply std/assembly/dataview.ts]: Not sure is it good idea. Currently `ArrayBufferView` and `DataView` structurally identical (have same layout and fields) but `ArrayBufferView` use default constructor which don't need in `DataView`

File: std/assembly/arraybuffer.ts (hunk 7 above)
Comment (dcodeIO) [reply std/assembly/arraybuffer.ts]: Hmm, since user code will solely access `.buffer`, I guess this depends on the use case? Like, if user code doesn't access `.buffer`, there'll be more `.dataStart`, but if it does, there might be a lot more `.buffer` at some point. Do you know if a load with an offset has implications other than code size?

Comment (MaxGraey) [reply std/assembly/arraybuffer.ts]: No it doesn't affect to codesize. Just on codegen:
```asm
mov eax, dword ptr [r15 + rdi] 
```
vs
```asm
mov eax, dword ptr [r15 + rdi + 4]
```

Comment (dcodeIO) [reply std/assembly/arraybuffer.ts]: I see, though I still wonder what's the advantage then. I'm a bit concerned here because not every existing program might use the loader, and do manual loads expecting that `buffer` and `dataStart` are at specific offsets instead, so these programs would break with the change.

Comment (MaxGraey) [reply std/assembly/arraybuffer.ts]: Ok, I'll revert this changes

Comment (MaxGraey) [reply std/assembly/arraybuffer.ts]: Done!

**Comment by dcodeIO** [comment]
LGTM :) Might be a good opportunity to also think about the `dataStart` name, which somewhat made sense while there was `data` and `dataLength`, but now is a bit unintuitive. I don't have a good idea for a better name, though.

**Comment by MaxGraey** [comment]
I guess better leave old name for `dataStart` because it's already used in exteranl third-party course code. Also it's unsafe and don't use in `safe`-context which I guess will be standard de-facto in future.

**Comment by dcodeIO** [comment]
There is a chance that removing `data` and  `dataLength` breaks these cases anyway, hence the suggestion. Isn't super important to rename it ofc, but if we want to do it I'd say that this is the PR - or we'll never do it :)

**Comment by MaxGraey** [comment]
Yeah. We could add something like:
```ts

@unsafe readonly rawptr: usize;

@inline function dataStart(): usize {
   ERROR("dataStart was deprecated. Please use rawPtr instead");
   return this.rawptr;
}

```

**Comment by MaxGraey** [comment]
@dcodeIO what about `rawBuffer`?

**Comment by dcodeIO** [comment]
Thinking about it a little more, it's "the unsafe pointer to the start of the data in the backing buffer", so `dataStart` isn't that bad maybe, with only `dataStartPtr` or maybe a short `unsafePtr` being more concise but not worth the trouble I guess. Let's just keep it :)

**Comment by dcodeIO** [comment]
Thanks!
